File: functions/flask-app.py
import os
import sys
import json
from flask import Flask, request

# Add project to path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Import our Flask app
from comiccaster.web_interface import app as flask_app

def handler(event, context):
    """
    Netlify Function handler for the Flask app.
    """
    # Parse request details from Netlify event
    method = event['httpMethod']
    path = event['path']
    
    # Remove function prefix if present
    if path.startswith('/.netlify/functions/flask-app'):
        path = path.replace('/.netlify/functions/flask-app', '')
    
    # Ensure path starts with a slash
    if not path:
        path = '/'
    
    # Set environment variables
    os.environ['SERVER_NAME'] = event['headers'].get('host', 'comiccaster.xyz')
    
    # Prepare WSGI environment
    environ = {
        'REQUEST_METHOD': method,
        'PATH_INFO': path,
        'QUERY_STRING': '',
        'SERVER_NAME': os.environ['SERVER_NAME'],
        'SERVER_PORT': '443',
        'HTTP_HOST': os.environ['SERVER_NAME'],
        'wsgi.url_scheme': 'https',
        'wsgi.input': '',
        'wsgi.errors': sys.stderr,
    }
    
    # Handle query parameters
    if event.get('queryStringParameters'):
        query_string = '&'.join([f"{k}={v}" for k, v in event['queryStringParameters'].items()])
        environ['QUERY_STRING'] = query_string
    
    # Handle headers
    for header, value in event.get('headers', {}).items():
        key = header.upper().replace('-', '_')
        if key not in ('CONTENT_TYPE', 'CONTENT_LENGTH'):
            key = 'HTTP_' + key
        environ[key] = value
    
    # Handle body
    if event.get('body'):
        environ['wsgi.input'] = event['body']
        environ['CONTENT_LENGTH'] = str(len(event['body']))
    
    # Capture the response
    response = {'statusCode': 200, 'headers': {}, 'body': ''}
    
    def start_response(status, response_headers, exc_info=None):
        status_code = int(status.split(' ')[0])
        response['statusCode'] = status_code
        
        for key, value in response_headers:
            response['headers'][key] = value
    
    # Call the Flask app
    output = flask_app(environ, start_response)
    response['body'] = ''.join([chunk.decode('utf-8') if isinstance(chunk, bytes) else chunk for chunk in output])
    
    return response 

File: functions/flask-app.js
const { spawn } = require('child_process');
const path = require('path');
const process = require('process');

// Set up proper Python environment
process.env.PYTHONPATH = process.cwd() + ':' + process.env.PYTHONPATH;

// Function to execute Flask app as a subprocess
exports.handler = async function(event, context) {
  // Configure Flask app through environment variables
  process.env.FLASK_ENV = 'production';
  process.env.SERVER_NAME = event.headers.host || 'comiccaster.xyz';
  
  // Parse the path - remove the function path prefix if present
  let requestPath = event.path;
  if (requestPath.startsWith('/.netlify/functions/flask-app')) {
    requestPath = requestPath.replace('/.netlify/functions/flask-app', '');
  }
  if (requestPath === '') {
    requestPath = '/';
  }
  
  console.log('Request path:', requestPath);
  
  // Prepare arguments for Flask subprocess
  const args = [
    '-m', 'flask',
    'routes',  // Just list routes for debugging
    '--no-debugger'
  ];

  return new Promise((resolve, reject) => {
    try {
      // Call the Flask app with the request details
      const flask = spawn('python', args, {
        env: process.env,
        cwd: process.cwd()
      });
      
      let stdout = '';
      let stderr = '';
      
      flask.stdout.on('data', (data) => {
        stdout += data.toString();
      });
      
      flask.stderr.on('data', (data) => {
        stderr += data.toString();
      });
      
      flask.on('close', (code) => {
        if (code !== 0) {
          console.error(`Flask process exited with code ${code}`);
          console.error('Stderr:', stderr);
          
          // Return a helpful error message
          resolve({
            statusCode: 500,
            headers: {
              'Content-Type': 'text/html',
            },
            body: `
              <html>
                <head>
                  <title>ComicCaster - Service Error</title>
                  <style>
                    body { font-family: system-ui, sans-serif; line-height: 1.5; padding: 2rem; background: #1a1a1a; color: #fff; }
                    .container { max-width: 800px; margin: 0 auto; }
                  </style>
                </head>
                <body>
                  <div class="container">
                    <h1>ComicCaster Service Error</h1>
                    <p>We're experiencing technical difficulties with our Flask application.</p>
                    <p>You can still access individual comic feeds directly:</p>
                    <ul>
                      <li><a href="/rss/calvinandhobbes" style="color: #3b82f6;">/rss/calvinandhobbes</a></li>
                      <li><a href="/rss/dilbert" style="color: #3b82f6;">/rss/dilbert</a></li>
                      <li><a href="/rss/garfield" style="color: #3b82f6;">/rss/garfield</a></li>
                    </ul>
                    <a href="/" style="display: inline-block; margin-top: 1rem; padding: 0.5rem 1rem; background: #0066cc; color: white; text-decoration: none; border-radius: 4px;">Try Again</a>
                    <div style="margin-top: 2rem; border-top: 1px solid #333; padding-top: 1rem;">
                      <p>Error details:</p>
                      <pre style="background: #333; padding: 1rem; overflow: auto; border-radius: 4px;">${stderr}</pre>
                    </div>
                  </div>
                </body>
              </html>
            `
          });
          return;
        }
        
        console.log('Flask routes:', stdout);
        
        // For now, return information about our Flask app
        resolve({
          statusCode: 200,
          headers: {
            'Content-Type': 'text/html',
          },
          body: `
            <html>
              <head>
                <title>ComicCaster - Redirecting</title>
                <meta http-equiv="refresh" content="0;url=/" />
                <style>
                  body { font-family: system-ui, sans-serif; line-height: 1.5; padding: 2rem; background: #1a1a1a; color: #fff; }
                  .container { max-width: 800px; margin: 0 auto; text-align: center; }
                </style>
              </head>
              <body>
                <div class="container">
                  <h1>ComicCaster</h1>
                  <p>Redirecting to main app...</p>
                  <p>If you are not redirected, <a href="/" style="color: #3b82f6;">click here</a>.</p>
                </div>
              </body>
            </html>
          `
        });
      });
    } catch (error) {
      console.error('Error running Flask:', error);
      resolve({
        statusCode: 500,
        body: `Server error: ${error.message}`
      });
    }
  });
}; 